test/distributed/tensor/test_attention.py

```diff
@@ -86,12 +107,89 @@ def test_ring_attention_sdpa(self) -> None:
                 "test_forward_only": [True, False],
                 "dispatch_mode": [
                     _DispatchMode.MONKEY_PATCH,
-                    _DispatchMode.TORCH_FUNCTION,
+                    _DispatchMode.MODULE_WRAPPER,
                 ],
             },
             self._test_ring_attention_sdpa,
         )
 
+    def _ring_attention_sdpa(
+        self,
+        cp_q: torch.Tensor,
+        cp_k: torch.Tensor,
+        cp_v: torch.Tensor,
+        *,
+        fn_eval: Callable,
+        mesh: DeviceMesh,
+        seq_dim: int,
+        is_causal: bool,
+        compiled: bool,
+        backend: SDPBackend,
+        rotater: _RotateMethod,
+        test_forward_only: bool,
+        dispatch_mode: _DispatchMode,
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        if dispatch_mode == _DispatchMode.MODULE_WRAPPER:
+            cp_plan = _ContextParallel(
+                seq_dim=seq_dim,
+                attention_type=_ContextParallel.AttentionType.SDPA,
+            )
+            attention = SDPAWrapper(compiled=compiled, backend=backend)
+            attention = parallelize_module(attention, mesh, cp_plan)
+        else:
+            # This won't work for MONKEY_PATCH.
+            attention = F.scaled_dot_product_attention
+
+        # Theoretically, context_parallel() should not be used to shard
+        # parameters because when require_grad is True, resize_ is not
+        # allowed. But requires_grad of cp_q, cp_k, and cp_v are False
+        # now. So we can just use context_parallel() to shard q, k, v.
+        # In reality, context_paralle() should be used to shard the input.
+        with context_parallel(
+            mesh, buffers=(cp_q, cp_k, cp_v), buffer_seq_dims=(seq_dim,) * 3
+        ):
+            # NOTE: This demonstrates that monkey patching is not fully reliable.
+            # If we use SDPAWrapper directly, the monkey patching dispatch mode
+            # does not function correctly. To ensure proper behavior,
+            # F.scaled_dot_product_attention must be referenced within the
+            # context_parallel() scope.
+            if dispatch_mode == _DispatchMode.MONKEY_PATCH:
+                attention = F.scaled_dot_product_attention
+                if compiled:
+                    attention = torch.compile(
+                        attention, fullgraph=True, backend="aot_eager"
+                    )
```

Comment:
IIUC this is expected because Module wrapper approach applies to modules while the monkey patch is globally replacing SDPA (i.e. users won't be able to use the original SDPA function). Therefore these 2 things are conflicting. 

Likely we should deprecate the monkey patch approach but the deprecation time is TBD.